src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -193,9 +252,31 @@ protected void handleMessageInternal(final Message<?> message) throws Exception
 			headers = new RecordHeaders();
 			this.headerMapper.fromHeaders(message.getHeaders(), headers);
 		}
-		ProducerRecord<K, V> producerRecord = new ProducerRecord<K, V>(topic, partitionId, timestamp, (K) messageKey,
-				payload, headers);
-		ListenableFuture<?> future = this.kafkaTemplate.send(producerRecord);
+		final ProducerRecord<K, V> producerRecord = new ProducerRecord<K, V>(topic, partitionId, timestamp,
+				(K) messageKey, payload, headers);
+		ListenableFuture<SendResult<K, V>> future = this.kafkaTemplate.send(producerRecord);
+		if (getSendFailureChannel() != null || getOutputChannel() != null) {
+			future.addCallback(new ListenableFutureCallback<SendResult<K, V>>() {
+
+				@Override
+				public void onSuccess(SendResult<K, V> result) {
+					if (getOutputChannel() != null) {
+						KafkaProducerMessageHandler.this.messagingTemplate.send(getOutputChannel(),
+								getMessageBuilderFactory().withPayload(result.getRecordMetadata()).build());
```

Comment:
Since this is async send, I think we still need some correlation with the `requestMessage`.
Maybe place this `RecordMetadata` to the `headers`?